core/indigo-core/common/math/algebra.h

```diff
@@ -822,5 +824,234 @@ namespace indigo
         float _d;
     };
 
+    inline bool isPointInPolygon(const Vec2f& p, const std::vector<Vec2f>& poly)
+    {
+        bool in = false;
+        for (size_t i = 0, n = poly.size(); i < n; ++i)
+        {
+            size_t j = (i + 1) % n;
+            bool intersect =
+                ((poly[i].y > p.y) != (poly[j].y > p.y)) && (p.x < (poly[j].x - poly[i].x) * (p.y - poly[i].y) / (poly[j].y - poly[i].y) + poly[i].x);
+            if (intersect)
+                in = !in;
+        }
+        return in;
+    }
+
+    inline std::vector<Vec2f> getPointsInsidePolygon(const std::vector<Vec2f>& polygon1, const std::vector<Vec2f>& polygon2)
+    {
+        std::vector<Vec2f> result;
+        result.reserve(polygon1.size());
+        std::copy_if(polygon1.begin(), polygon1.end(), std::back_inserter(result), [&](auto& p) { return isPointInPolygon(p, polygon2); });
+        return result;
+    }
+
+    inline bool isPointInConvexPolygon(const Vec2f& p, const std::vector<Vec2f>& poly)
+    {
+        auto cross = [](const Vec2f& a, const Vec2f& b, const Vec2f& c) { return (b.x - a.x) * (c.y - a.y) - (b.y - a.y) * (c.x - a.x); };
+        bool sign = cross(poly.back(), poly[0], p) < 0;
+        for (size_t i = 0, n = poly.size(); i < n; ++i)
+            if ((cross(poly[i], poly[(i + 1) % n], p) < 0) != sign)
+                return false;
+        return true;
+    }
+
+    inline std::vector<Vec2f> getPointsInsideConvexPolygon(const std::vector<Vec2f>& polygon1, const std::vector<Vec2f>& polygon2)
+    {
+        std::vector<Vec2f> result;
+        result.reserve(polygon1.size());
+        std::copy_if(polygon1.begin(), polygon1.end(), std::back_inserter(result), [&](auto& p) { return isPointInConvexPolygon(p, polygon2); });
+        return result;
+    }
+
+    inline float convexPolygonArea(const std::vector<Vec2f>& poly)
+    {
+        float area = 0.0f;
+        size_t n = poly.size();
+        for (size_t i = 0; i < n; ++i)
+        {
+            const Vec2f& p1 = poly[i];
+            const Vec2f& p2 = poly[(i + 1) % n];
+            area += Vec2f::cross(p1, p2);
+        }
+        return std::abs(area) * 0.5f;
+    }
+
+    inline bool isInside(const Vec2f& edgeStart, const Vec2f& edgeEnd, const Vec2f& point)
```

Comment:
I think it should be named isNormalNegative, and comment about "check sign of normal to (edgeStart,edgeEnd),(edgeStart,point)" will be usefull

Reply:
isInside has been deleted